MT3-657 #ready-for-test Add change links to Review sections

Gweaton opened the pull request.

erbridge commented on helpers/useReviewSectionRows.tsx:
```diff
@@ -400,6 +417,14 @@ const useReviewSectionRows = <
                   ))}
                 </div>
+                <div>
+                  <a
+                    className="govuk-link"
+                    href={`${changeLink.as.pathname}?review=true`}
+                  >
+                    Change
+                  </a>
```
This should be a `Link` tag (from `lbh-frontend-react`). This is so that it hooks in to the router properly so it will work offline.

erbridge commented on helpers/useReviewSectionRows.tsx:
```diff
@@ -377,6 +385,15 @@ const useReviewSectionRows = <
+          const changeLink = urlsForRouter(
+            router,
+            urlObjectForSlug(router, changeSlug)
+          );
```
Let's extend this method to accept an object of query parameters, so we don't have to deal with query parameters ad-hoc.

Gweaton replied:
I've updated this - for some reason I haven't had to modify `urlsForRouter`, after inspecting the query inside it, it seems to be passing it around correctly, unless I'm misunderstanding how it's supposed to...

erbridge replied:
No, I could believe that.

Gweaton committed:
Add change links to Review sections

Once taken back to the step, clicking "Save & Continue" will take you
back to the Review page

Extend `urlObjectForSlug` to accept query params

erbridge commented on helpers/useReviewSectionRows.tsx:
```diff
@@ -400,6 +418,13 @@ const useReviewSectionRows = <
                   ))}
                 </div>
+                <div>
+                  <NextLink href={changeLink.href} as={changeLink.as}>
+                    <a className="govuk-link" href={changeLink.as.pathname}>
+                      Change
+                    </a>
+                  </NextLink>
```
Ok, so having said to use `NextLink`, I think we actually want to use `Link` from `lbh-frontend-react`, which means constructing the target URL including query parameters to pass in as its `href`. `Link` is set up to do this work properly (the href here is missing the query parameter), with the correct styling via `_app.tsxL20-53`.

erbridge commented on components/InternalLink.tsx:
```diff
@@ -0,0 +1,52 @@
+  const { as } = url;
+
+  return (
+    <Link href={combinePathAndQueryString(as)}>
+      <a>Change</a>
+    </Link>
```
```suggestion
    <Link href={combinePathAndQueryString(as)}>
      {children}
    </Link>
```

Where `children` in this case is `"Change"`. `Link` already includes the `<a>`.

erbridge commented on helpers/useReviewSectionRows.tsx:
```diff
@@ -1,4 +1,4 @@
-import NextLink from "next/link";
+import InternalLink from "components/InternalLink";
```
```suggestion
import InternalLink from "../components/InternalLink";
```

erbridge commented on helpers/useReviewSectionRows.tsx:
```diff
@@ -14,14 +16,18 @@ import {
+import PageSlugs, {
+  repeatingStepSlugs,
+  urlObjectForSlug,
+} from "steps/PageSlugs";
```
```suggestion
} from "../steps/PageSlugs";
```

erbridge commented on helpers/useReviewSectionRows.tsx:
```diff
@@ -400,6 +418,11 @@ const useReviewSectionRows = <
                 </div>
+                <div className="change-link">
+                  <InternalLink url={changeLink.as}>
+                    <a>Change</a>
```
```suggestion
                    Change
```

erbridge reviewed (commented):
Very minor points.

erbridge commented on components/InternalLink.tsx:
```diff
@@ -0,0 +1,38 @@
+InternalLink.propTypes = {
+  url: PropTypes.shape({
+    pathname: PropTypes.string.isRequired,
+    query: PropTypes.shape({
+      s: PropTypes.string,
+    }),
```
```suggestion
    query: PropTypes.object,
```

We don't know anything about the internals of the query, so this is fine.

erbridge reviewed (approved):
Minor prop type issue, but otherwise approved in advance.
